Base contact by function

bwrsandman opened the pull request:
- [ADD] added base_contact_function module: It allow to manage contacts by function"
- [ADD] Added base_contact_function_partner_firstname module: It allows to add firstname before name.Name is considered like the lastname and we have firstname and lastname on partner view and contact view."
  The base_contact_function_partner_firstname module depends on partner_firstname and base_contact_function module.
## Previous reviews:

@StefanRijnhart

> I have a hard time trying to find out what this module actually does. The manifest description is very terse. The name of the module reminds me of base_contact, which allowed multiple connections between the same person and a company. Is this module similar? The description should at least explain what 'institutions' are (a concept that this module introduces).
> 
> Terminology is off at times, it seems. l.130: 'acronym': is this abbreviation, or maybe stick to the standard 'code' field name and label?
> 
> Organism: I'm guessing you mean organization. But in any case, changing the 'company' terminology should be split off as a separate module.
> 
> The translations seem to contain obsolete terms, e.g. l.241 and other occurrences of 'passport'.

@bwrsandman

> This is for cases where one partner (Mike Fletcher President of Agrolait) can be part of another company (Mike Fletcher VP of AsusTEK) in a different function.
> 
> The module helps track these functions.

ehdem committed:
[ADD] base_contact_function_partner_firstname

Add firstname before name.
Name is considered like the lastname and we have firstname and lastname on
partner view and contact view.

nbessi commented on base_contact_by_functions/__init__.py:
```diff
@@ -0,0 +1,27 @@
+from . import functions
+from . import res_institutions
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
vim line

nbessi commented on base_contact_by_functions/__openerp__.py:
```diff
@@ -0,0 +1,56 @@
+Contacts by Functions
+=====================
+
+This module allows you to manage contacts by functions
```
This should be completed with how to use it

bwrsandman replied:
ping @giotta

nbessi commented on base_contact_by_functions/__openerp__.py:
```diff
@@ -0,0 +1,56 @@
+    'active': False,
+}
+
+# vim:expandtab:smartindent:tabstop=4:softtabstop=4:shiftwidth=4:
```
vim line

nbessi commented on base_contact_by_functions/functions.py:
```diff
@@ -0,0 +1,52 @@
+        ),
+        'acronym': fields.char(
+            'Acronym',
+            size=50,
```
maybe add a select=True on acronym?

bwrsandman replied:
Done

nbessi commented on base_contact_by_functions/functions_view.xml:
```diff
@@ -0,0 +1,38 @@
+      <field name="model">functions</field>
+      <field name="type">form</field>
+      <field name="arch" type="xml">
+        <form string="FunctionsForm" version="7.0" >
```
FunctionsForm -> Functions. It think that it is shown in View

nbessi commented:
:+1:

gurneyalex commented on base_contact_by_functions/functions.py:
```diff
@@ -0,0 +1,53 @@
+
+class functions(orm.Model):
+    _description = 'Functions'
+    _name = 'functions'
```
I'm concerned this may be a bit too generic and confusing (function fields...). Also the convention nowadays is to use a singular form for Model `_name`. 

I suggest `"res.partner.function"` for `_name`.

gurneyalex commented on base_contact_by_functions/functions.py:
```diff
@@ -0,0 +1,53 @@
+    _description = 'Functions'
+    _name = 'functions'
+    _columns = {
+        'name': fields.char(
```
should this field be translatable?

gurneyalex commented on base_contact_by_functions/functions.py:
```diff
@@ -0,0 +1,53 @@
+            select=True,
+            help='Name of function.',
+        ),
+        'acronym': fields.char(
```
should this field be translatable?

gurneyalex commented on base_contact_by_functions/res_institutions.py:
```diff
@@ -0,0 +1,70 @@
+
+class res_institutions_functions(orm.Model):
+    _description = 'Institutions functions'
+    _name = 'res.institution.functions'
```
same remark as above about singular form for `_name`

gurneyalex commented on base_contact_by_functions/res_partner.py:
```diff
@@ -0,0 +1,135 @@
+    Adds this class to link category and functions
+    """
+    _description = 'Categories functions'
+    _name = 'res.category.functions'
```
same remark as above regarding singular form for `_name`

gurneyalex commented on base_contact_by_functions_partner_firstname/tests/test_partner.py:
```diff
@@ -0,0 +1,51 @@
+
+    def test_create_partner(self):
+        cr, uid, vals, context = self.cr, self.uid, self.vals, self.context
+        self.partner_model.create(cr, uid, vals, context=context)
```
could we have a couple of assertions checking that both function and first name are working properly?

gurneyalex commented on base_contact_by_functions/tests/test_institutions.py:
```diff
@@ -0,0 +1,103 @@
+        Setting up institution.
+        """
+        # Default test values
+        self.vals = {'name': 'This is a test institution name',
```
please update the test with a unicode string for name, containing non ascii characters.

gurneyalex commented on base_contact_by_functions/tests/test_institutions.py:
```diff
@@ -0,0 +1,103 @@
+            else:
+                self.assertEquals(str(self.vals[field]), str(val),
+                                  "Values for %s don't match: (%s != %s)" %
+                                  (field, str(self.vals[field]), str(val)))
```
str(val) is wrong, it will crash with a unicode string. Please fix the test to use unicode strings everywhere.

gurneyalex commented on base_contact_by_functions/tests/test_institutions.py:
```diff
@@ -0,0 +1,103 @@
+            else:
+                self.assertNotEqual(str(self.vals[field]), str(val),
+                                    "Values for %s don't match: (%s != %s)" %
+                                    (field, str(self.vals[field]), str(val)))
```
use Unicode strings

gurneyalex commented on base_contact_by_functions/tests/test_partner_category.py:
```diff
@@ -0,0 +1,112 @@
+            else:
+                self.assertEquals(str(self.vals[field]), str(val),
+                                  "Values for %s don't match: (%s != %s)" %
+                                  (field, str(self.vals[field]), str(val)))
```
use unicode strings

gurneyalex commented on base_contact_by_functions/tests/test_partner_category.py:
```diff
@@ -0,0 +1,112 @@
+        Setting up partner_category.
+        """
+        # Default test values
+        self.vals = {'name': 'This is a test partner_category name',
```
please update the test with a unicode string with non ascii chars in name.

gurneyalex commented on base_contact_by_functions/tests/test_partner_category.py:
```diff
@@ -0,0 +1,112 @@
+            else:
+                self.assertNotEqual(str(self.vals[field]), str(val),
+                                    "Values for %s don't match: (%s != %s)" %
+                                    (field, str(self.vals[field]), str(val)))
```
use unicode strings

gurneyalex commented on base_contact_by_functions_partner_firstname/tests/test_partner.py:
```diff
@@ -0,0 +1,51 @@
+            'title': 1,
+        }, context=self.context)
+        self.vals = {
+            'name': 'test_partner',
```
please test with a unicode string containing non-ascii characters for name (e.g. u"André Barnabé")

gurneyalex commented on base_contact_by_functions/tests/test_functions.py:
```diff
@@ -0,0 +1,101 @@
+            else:
+                self.assertEquals(str(self.vals[field]), str(val),
+                                  "Values for %s don't match: (%s != %s)" %
+                                  (field, str(self.vals[field]), str(val)))
```
don't cast potentially Unicode objects to strings. This will crash if there is a non-ascii character.

gurneyalex commented on base_contact_by_functions/tests/test_functions.py:
```diff
@@ -0,0 +1,101 @@
+            else:
+                self.assertNotEqual(str(self.vals[field]), str(val),
+                                    "Values for %s don't match: (%s != %s)" %
+                                    (field, str(self.vals[field]), str(val)))
```
fix str() call on unicode object.

gurneyalex commented on base_contact_function_partner_firstname/tests/test_partner.py:
```diff
@@ -0,0 +1,57 @@
+        self.context = self.user_model.context_get(self.cr, self.uid)
+        # Create values for test, contact also created
+        contact_id = self.partner_model.create(self.cr, self.uid, {
+            'name': 'Astérix',
```
s/'Astérix'/u'Astérix'/

bwrsandman replied:
Why? This case catches more unicode errors.

gurneyalex replied:
On 03/09/2014 13:52, Sandy wrote:

> In base_contact_function_partner_firstname/tests/test_partner.py:
> 
> > +
> > +class test_partner_contact_id(TransactionCase):
> > +
> > -    def setUp(self):
> > -        super(test_partner_contact_id, self).setUp()
> > -        # Clean up registries
> > -        self.registry('ir.model').clear_caches()
> > -        self.registry('ir.model.data').clear_caches()
> > -        # Get registries
> > -        self.user_model = self.registry("res.users")
> > -        self.partner_model = self.registry("res.partner")
> > -        # Get context
> > -        self.context = self.user_model.context_get(self.cr, self.uid)
> > -        # Create values for test, contact also created
> > -        contact_id = self.partner_model.create(self.cr, self.uid, {
> > -            'name': 'Astérix',
> 
> Why? This case catches more unicode errors.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/OCA/partner-contact/pull/20/files#r17043924.
> 
> I won't fight over this. It's just a better practice IMO, always using
> unicode strings where the API mandates unicode, rather than utf-8
> encoded string objects. This makes your source code the same regardless
> of the encoding declaration at the top of the file.

## 

Alexandre Fayolle
Chef de Projet
Tel : + 33 (0)4 79 26 57 94

Camptocamp France SAS
Savoie Technolac, BP 352
73377 Le Bourget du Lac Cedex
http://www.camptocamp.com

bwrsandman replied:
I would agree with you if it was in the code, but this is in the unittests

gurneyalex commented on base_contact_function_partner_firstname/tests/test_partner.py:
```diff
@@ -0,0 +1,57 @@
+            'title': 1,
+        }, context=self.context)
+        self.vals = {
+            'name': 'Obélix',
```
same thing here

gurneyalex commented on base_contact_function/tests/test_functions.py:
```diff
@@ -0,0 +1,103 @@
+        # Default test values
+        self.vals = {
+            'name': 'This is a test function name with uniödé',
+            'acronym': 'This is a test function acronym with uniödé',
```
you need the u"" string form for the values. I agree this work because odoo does some processing to try to process utf-8 encoded string objects (what you are passing here) and decode the utf-8, but you really should be using unicode here.

bwrsandman replied:
I am intentionally not using unicode to catch more encoding errors. I.e. using `unicode` instead of `ustr`

gurneyalex commented:
:+1:

hbrunn commented on base_contact_function/res_partner.py:
```diff
@@ -0,0 +1,97 @@
+        # Replace company by Organism
+        'use_parent_address': fields.boolean(
+            'Use Organism Address',
+            help="Select this if you want to set organism's address "
```
Does 'organism' have a meaning that's not mentioned here: http://www.oxforddictionaries.com/definition/english/organism ?

hbrunn commented:
Please be much more extensive in the addon's description. Especially, you should mention here how institutions, functions and categories work together.

Commit:
[ADD] base_contact_function module

Allow to manage contacts by function

Commit:
Extend base_contact_function description

bwrsandman commented:
@hbrunn addon's description provided by @giotta

bwrsandman commented:
Inconsistent fail between Odoo and OCB

Commit:
Add optional web_m2x_options options

Functions for contacts of companies should be created in the Functions
Tab. Functions created straight in the Contact form will not end up in the
Functions Tab.

Does not change behaviour on systems without web_m2x_options

bwrsandman committed:
Fix pylint

hbrunn commented on base_contact_function/res_partner_category.py:
```diff
@@ -0,0 +1,57 @@
+            name = name.split(' / ')[-1]
+            ids = self.search(cr, user, [('name', operator, name)] + args,
+                              limit=limit, context=context)
+            if ids:
```
this whole section doesn't respect the limit parameter, that will yield too many results when there are a lot of subcategories

hbrunn commented on base_contact_function/res_partner_category.py:
```diff
@@ -0,0 +1,57 @@
+                    cr, user, [('parent_id', 'child_of', ids)],
+                    context=context)
+                if child_ids:
+                    ids.extend(child_ids)
```
the childof operator also returns the id itself, so add them only if not included already

hbrunn commented:
@bwrsandman the difference is caused by ocb using browse() in name_get and odoo using read(). The latter discards double ids and gives a random(?) order, the former delivers all ids in the order of the id list.

So to fix the test, we shouldn't pass double ids to name_get as pointed out in my comment above, and we should only check for set equality (= no order)

Commit:
Respect limits and remove duplicates

bwrsandman commented:
@hbrunn I have fixed the limits and the duplicates problem.

I don't understand this issue with id order, though. In this implementation, I used `set` to remove duplicates, this, however is unordered. Is this good?

hbrunn commented:
I personally think that every name_get should return its list sorted alphabetically, but not all name_gets do that.

As for the issue at hand: The test fails because it checks for list equality (https://github.com/OCA/partner-contact/pull/20/files#diff-482719082e50caa32544883db16a8259R156) and there, order matters. Given that I don't think we care about the lists' orders, I think you'd better use assertItemsEqual: https://docs.python.org/2/library/unittest.html#unittest.TestCase.assertItemsEqual

Commit:
Don't test list order

bwrsandman commented:
Thanks @hbrunn.

As for the order, I thought odoo relied on the defined `_order`.

bwrsandman commented:
>  The Travis CI build passed

:raised_hands:

hbrunn commented:
yay! Concerning `_order`: `search()` uses it, and `browse()` returns records in the order it got in the id list, but `read()` returns the records as they come from the database.

Anyway: :+1:

nbessi commented:
:+1:
